azure/compute/src/main/java/org/jclouds/azurecompute/compute/functions/VirtualMachineToNodeMetadata.java

```diff
@@ -0,0 +1,109 @@
+import java.util.Set;
+
+import javax.annotation.Resource;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import com.google.common.collect.Collections2;
+import com.google.common.collect.ImmutableList;
+import com.microsoft.windowsazure.management.compute.models.DeploymentStatus;
+import com.microsoft.windowsazure.management.compute.models.VirtualIPAddress;
+import org.jclouds.azurecompute.domain.VirtualMachine;
+import org.jclouds.collect.Memoized;
+import org.jclouds.compute.domain.Hardware;
+import org.jclouds.compute.domain.Image;
+import org.jclouds.compute.domain.NodeMetadata;
+import org.jclouds.compute.domain.NodeMetadataBuilder;
+import org.jclouds.compute.domain.NodeMetadata.Status;
+import org.jclouds.compute.functions.GroupNamingConvention;
+import org.jclouds.compute.predicates.HardwarePredicates;
+import org.jclouds.compute.predicates.ImagePredicates;
+import org.jclouds.domain.Location;
+import org.jclouds.logging.Logger;
+
+import com.google.common.base.Function;
+import com.google.common.base.Supplier;
+
+@Singleton
+public class VirtualMachineToNodeMetadata implements Function<VirtualMachine, NodeMetadata> {
+
+   @Resource
+   protected Logger logger = Logger.NULL;
+   private final Supplier<Set<? extends Image>> images;
+   private final Supplier<Set<? extends Hardware>> hardwares;
+   private final Supplier<Set<? extends Location>> locations;
+   private final GroupNamingConvention nodeNamingConvention;
+   private final Map<DeploymentStatus, Status> deploymentStatusToStatus;
+   private final Map<String, Status> instanceStatusToStatus;
+
+   private static final Function<VirtualIPAddress, String> toIpAddress = new Function<VirtualIPAddress, String>() {
+       @Override
+       public String apply(VirtualIPAddress input) {
+           return input.getAddress() == null ? null : input.getAddress().getHostAddress();
+       }
+   };
+
+   @Inject
+   VirtualMachineToNodeMetadata(@Memoized Supplier<Set<? extends Image>> images, @Memoized Supplier<Set<? extends Hardware>> hardwares,
+                                Supplier<Set<? extends Location>> locations,
+                                GroupNamingConvention.Factory namingConvention,
+                                Map<DeploymentStatus, Status> deploymentStatusToStatus, Map<String, Status> instanceStatusToStatus) {
+      this.nodeNamingConvention = checkNotNull(namingConvention, "namingConvention").createWithoutPrefix();
+      this.images = checkNotNull(images, "images");
+      this.hardwares = checkNotNull(hardwares, "hardwares");
+      this.locations = checkNotNull(locations, "locations");
+      this.deploymentStatusToStatus = checkNotNull(deploymentStatusToStatus);
+      this.instanceStatusToStatus = instanceStatusToStatus;
+   }
+
+   @Override
+   public NodeMetadata apply(VirtualMachine from) {
+      NodeMetadataBuilder builder = new NodeMetadataBuilder();
+      builder.ids(from.getServiceName() + ":" + from.getRoleName());
```

Comment:
Same comment about the image IDs. You might want to set separately the `id` and the `providerId`.